### PR TITLE
perlPackages.ModuleBuild and .FileBaseDir: fix cross-compilation WIP

### LINKED ISSUE #66741: perlPackages.ModuleBuild (a dependency of git) fails to cross-compile
`perlPackages.ModuleBuild` fails to cross-compile, which prevents `git` from cross-compiling (unless `perlSupport = false`). I don't have much experience with Perl, but my understanding is that `Scalar::Util` is trying to load an XS module. I'm not sure how to work around this.

```
unpacking sources
unpacking source archive /nix/store/jwpgfd4pziwrrx2hywr9ckf9xil7aizc-Module-Build-0.4229.tar.gz
source root is Module-Build-0.4229
setting SOURCE_DATE_EPOCH to timestamp 1555368693 of file Module-Build-0.4229/t/xs.t
patching sources
updateAutotoolsGnuConfigScriptsPhase
configuring
patching ./t/properties/share_dir.t...
patching ./t/compat/exit.t...
patching ./t/xs.t...
patching ./t/write_default_maniskip.t...
patching ./t/versions.t...
patching ./t/use_tap_harness.t...
patching ./t/unit_run_test_harness.t...
patching ./t/tilde.t...
patching ./t/test_types.t...
patching ./t/test_type.t...
patching ./t/test_reqs.t...
patching ./t/test_file_exts.t...
patching ./t/signature.t...
patching ./t/script_dist.t...
patching ./t/runthrough.t...
patching ./t/ppm.t...
patching ./t/pod_parser.t...
patching ./t/parents.t...
patching ./t/par.t...
patching ./t/notes.t...
patching ./t/new_from_context.t...
patching ./t/mymeta.t...
patching ./t/metadata2.t...
patching ./t/metadata.t...
patching ./t/manifypods.t...
patching ./t/install_extra_target.t...
patching ./t/install.t...
patching ./t/help.t...
patching ./t/files.t...
patching ./t/extend.t...
patching ./t/ext.t...
patching ./t/destinations.t...
patching ./t/debug.t...
patching ./t/compat.t...
patching ./t/basic.t...
patching ./t/add_property_hash.t...
patching ./t/add_property_array.t...
patching ./t/add_property.t...
patching ./t/PL_files.t...
patching ./bin/config_data...
# running Build.PL --prefix /nix/store/9lm1ijj01kmimdplwrs77faflwzxjir3-perl5.28.2-Module-Build-0.4229-armv6l-unknown-linux-gnueabihf --installdirs site --config perl=/nix/store/dv4w3l6jclic0wqbkpwv6m79wbfkvk9n-perl-5.28.2-armv6l-unknown-linux-gnueabihf-dev/bin/perl --fullperl "/nix/store/by0xi2dd0aha8armhbkd2a3zdqzrjnmi-perl-5.28.2/bin/perl"
Checking prerequisites...
  test_requires:
    !  Couldn't find a $VERSION in prerequisite File::Temp

Checking optional features...
inc_bundling_support....disabled
  requires:
    ! inc::latest is not installed

ERRORS/WARNINGS FOUND IN PREREQUISITES.  You may wish to install the versions
of the modules indicated above before proceeding with this installation

Attempt to reload Scalar/Util.pm aborted.
Compilation failed in require at /nix/store/81lwy2awhy4fqxxiyqcc9jflslrzl2bm-perl-5.28.2-armv6l-unknown-linux-gnueabihf/lib/perl5/5.28.2/armv6l-linux/Data/Dumper.pm line 312.
Couldn't run Build.PL: No such file or directory at lib/Module/Build/Compat.pm line 341.
```

Maintainer information:
```yaml
attribute:
  - perlPackages.ModuleBuild
  - git
```

cc @shlevy @volth because you've worked on Perl cross-compiling

### PR DESCRIPTION
**Pull request by ck3d**
###### Description of changes

This PR fixes #66741, the cross compilation for Perl module ModuleBuild and FileBaseDir. Following command will succeed:

```sh
nix-build -A pkgsCross.aarch64-multiplatform.perlPackages.ModuleBuild
nix-build -A pkgsCross.aarch64-multiplatform.perlPackages.FileBaseDir
```

A cross compiled Perl module gets pkgs.perl.mini passed instead of pkgs.perl. The mini do not allow dynamic loading, which causes errors during build.

definition of perl.mini:
https://github.com/NixOS/nixpkgs/blob/master/pkgs/development/interpreters/perl/default.nix#L34

usage of perl.mini:
https://github.com/NixOS/nixpkgs/blob/master/pkgs/development/perl-modules/generic/default.nix#L46

Following issues suffer from the same problem and can hopefully fixed in a similar way:

- https://github.com/NixOS/nixpkgs/issues/198548
- https://github.com/NixOS/nixpkgs/issues/142122

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.05 Release Notes (or backporting 22.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2305-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/top-level/perl-packages.nix [hunk 1]
```diff
@@ -15233,6 +15233,10 @@ let
       url = "mirror://cpan/authors/id/L/LE/LEONT/Module-Build-0.4231.tar.gz";
       hash = "sha256-fg9MaSwXQMGshOoU1+o9i8eYsvsmwJh3Ip4E9DCytxc=";
     };
+    postPatch = lib.optionalString (stdenv.hostPlatform != stdenv.buildPlatform) ''
+      # use fullperl since default perl do not support dynamic linking
+      export PATH="$fullperl/bin:$PATH"
```
Comment (NickCao) [inline pkgs/top-level/perl-packages.nix:15238]: use `addToSearchPath PATH "$fullperl/bin"`

Comment (ck3d) [reply pkgs/top-level/perl-packages.nix:15238]: Your proposal doesn't work. The function addToSearchPath adds dir at the end off and not at the beginning.

**Comment by amjoseph-nixpkgs** [comment]
@ofborg build pkgsCross.aarch64-multiplatform.xdg-utils

**Comment by amjoseph-nixpkgs** [comment]
[No luck](https://logs.nix.ci/?key=nixos/nixpkgs.222427&attempt_id=607bf5fb-12dd-442b-8731-2e3b10a0f6e5)

**Comment by ck3d** [comment]
I added the same patch to FileBaseDir. You can now cross-compile xdg-utils

**Comment by NickCao** [comment]
Basically the same fix, but in another approach, is being carried out in https://github.com/NixOS/nixpkgs/pull/225086, may you also take a look?

**Comment by ck3d** [comment]
Thanks, looks better than my approach. I will update this PR.

**Review by apfelkuchen6 — APPROVED** [review]
LGTM.

Since this causes 0 rebuilds, this can go to master directly.

**Comment by Artturin** [comment]
Isn't the reason why we use miniperl is that the normal perl doesn't cross compile stuff?

**Comment by apfelkuchen6** [comment]
Oh, shit:
```
➜ find /nix/store/l5dn2khpgsf74ai5kjr0qmc3xwdx3n8f-perl5.36.0-Module-Build-0.4231-aarch64-unknown-linux-gnu      
[...]
/nix/store/l5dn2khpgsf74ai5kjr0qmc3xwdx3n8f-perl5.36.0-Module-Build-0.4231-aarch64-unknown-linux-gnu/lib/perl5/site_perl/5.36.0/Module/Build/PodParser.pm
/nix/store/l5dn2khpgsf74ai5kjr0qmc3xwdx3n8f-perl5.36.0-Module-Build-0.4231-aarch64-unknown-linux-gnu/lib/perl5/site_perl/5.36.0/x86_64-linux-thread-multi
/nix/store/l5dn2khpgsf74ai5kjr0qmc3xwdx3n8f-perl5.36.0-Module-Build-0.4231-aarch64-unknown-linux-gnu/lib/perl5/site_perl/5.36.0/x86_64-linux-thread-multi/auto
/nix/store/l5dn2khpgsf74ai5kjr0qmc3xwdx3n8f-perl5.36.0-Module-Build-0.4231-aarch64-unknown-linux-gnu/lib/perl5/site_perl/5.36.0/x86_64-linux-thread-multi/auto/Module
/nix/store/l5dn2khpgsf74ai5kjr0qmc3xwdx3n8f-perl5.36.0-Module-Build-0.4231-aarch64-unknown-linux-gnu/lib/perl5/site_perl/5.36.0/x86_64-linux-thread-multi/auto/Module/Build
/nix/store/l5dn2khpgsf74ai5kjr0qmc3xwdx3n8f-perl5.36.0-Module-Build-0.4231-aarch64-unknown-linux-gnu/lib/perl5/site_perl/5.36.0/x86_64-linux-thread-multi/auto/Module/Build/.packlist
```

**Comment by ck3d** [comment]
The differences between perl and perl.mini in case of cross compilation are:

1. Scrips in bin/ point to the wrong perl interpreter. These scripts will not run on the target.
2. Target specific module directory is wrong. The directory name should be aarch64-linux/, but it is named x86_64-linux-thread-multi/.

I will mark this PR as WIP as long we did not fix these issues.

**Pull request closed without merging**